File: backend/app/controllers/tasks.py
from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session

from app.core.dependencies import get_current_user
from app.database import get_db
from app.models.task import TaskStatus
from app.models.user import User
from app.schemas.task import PaginatedTasks, TaskCreate, TaskOut, TaskUpdate
from app.services import task_service

router = APIRouter(prefix="/tasks", tags=["tasks"])


@router.get(
    "",
    response_model=PaginatedTasks,
    summary="List tasks",
    description=(
        "Returns a paginated list of tasks. Optionally filter by `status` and/or `assigned_user_id`. "
        "Results are ordered by creation date descending. Deleted tasks are never returned."
    ),
    responses={401: {"description": "Missing or invalid token"}},
)
def list_tasks(
    page: int = Query(1, ge=1, description="Page number, starting at 1."),
    page_size: int = Query(20, ge=1, le=100, description="Number of results per page (max 100)."),
    status_filter: TaskStatus | None = Query(None, alias="status", description="Filter by task status."),
    assigned_user_id: int | None = Query(None, description="Filter by assigned user ID."),
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    return task_service.get_tasks(db, page, page_size, status_filter, assigned_user_id)


@router.post(
    "",
    response_model=TaskOut,
    status_code=status.HTTP_201_CREATED,
    summary="Create task",
    description="Creates a new task. The authenticated user becomes the creator. `title` is required. `status` defaults to `Todo`.",
    responses={
        401: {"description": "Missing or invalid token"},
        422: {"description": "Validation error — title is empty or assigned user does not exist"},
    },
)
def create_task(
    data: TaskCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return task_service.create_task(data, current_user, db)


@router.put(
    "/{task_id}",
    response_model=TaskOut,
    summary="Update task",
    description="Partially updates a task. Only the task creator or the assigned user can edit. Only fields included in the request body are changed.",
    responses={
        401: {"description": "Missing or invalid token"},
        403: {"description": "Not the creator or assignee"},
        404: {"description": "Task not found"},
        422: {"description": "Validation error — title is empty or assigned user does not exist"},
    },
)
def update_task(
    task_id: int,
    data: TaskUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return task_service.update_task(task_id, data, current_user, db)


@router.post(
    "/{task_id}/claim",
    response_model=TaskOut,
    summary="Claim task",
    description="Assigns the authenticated user to an unassigned task. Fails if the task already has an assignee.",
    responses={
        401: {"description": "Missing or invalid token"},
        404: {"description": "Task not found"},
        409: {"description": "Task is already assigned"},
    },
)
def claim_task(
    task_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return task_service.claim_task(task_id, current_user, db)


@router.delete(
    "/{task_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete task",
    description="Soft-deletes a task. Only the task creator can delete.",
    responses={
        401: {"description": "Missing or invalid token"},
        403: {"description": "Not the task creator"},
        404: {"description": "Task not found"},
    },
)
def delete_task(
    task_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    task_service.delete_task(task_id, current_user, db)

File: frontend/src/stores/tasks.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import api from '@/services/api'

export type TaskStatus = 'Todo' | 'In Progress' | 'Done'

export interface User {
  id: number
  name: string
  email: string
}

export interface Task {
  id: number
  title: string
  description: string | null
  status: TaskStatus
  assigned_user_id: number | null
  assigned_user: User | null
  created_by_id: number | null
  created_by: User | null
  created_at: string
  updated_at: string
}

export interface TaskFilters {
  status?: TaskStatus | null
  assigned_user_id?: number | null
}

export const useTasksStore = defineStore('tasks', () => {
  const tasks = ref<Task[]>([])
  const total = ref(0)
  const totalPages = ref(1)
  const page = ref(1)
  const pageSize = ref(20)
  const loading = ref(false)
  const filters = ref<TaskFilters>({})

  const users = ref<User[]>([])

  async function fetchTasks() {
    loading.value = true
    try {
      const params: Record<string, any> = { page: page.value, page_size: pageSize.value }
      if (filters.value.status) params.status = filters.value.status
      if (filters.value.assigned_user_id) params.assigned_user_id = filters.value.assigned_user_id

      const res = await api.get('/tasks', { params })
      tasks.value = res.data.items
      total.value = res.data.total
      totalPages.value = res.data.total_pages
    } finally {
      loading.value = false
    }
  }

  async function fetchUsers() {
    const res = await api.get('/users')
    users.value = res.data
  }

  async function createTask(data: { title: string; description?: string | null; status: TaskStatus; assigned_user_id?: number | null }) {
    await api.post('/tasks', data)
    await fetchTasks()
  }

  async function updateTask(id: number, data: Partial<{ title: string; description: string | null; status: TaskStatus; assigned_user_id: number | null }>) {
    await api.put(`/tasks/${id}`, data)
    await fetchTasks()
  }

  async function claimTask(id: number) {
    await api.post(`/tasks/${id}/claim`)
    await fetchTasks()
  }

  async function deleteTask(id: number) {
    await api.delete(`/tasks/${id}`)
    await fetchTasks()
  }

  function setPage(p: number) {
    page.value = p
    fetchTasks()
  }

  function setFilters(f: TaskFilters) {
    filters.value = f
    page.value = 1
    fetchTasks()
  }

  return { tasks, total, totalPages, page, pageSize, loading, filters, users, fetchTasks, fetchUsers, createTask, updateTask, claimTask, deleteTask, setPage, setFilters }
})
